AKS SPEC

dvdplm opened the pull request:
Add a `SPEC.md`. Let's use this PR to work on the spec. Whether we eventually merge it or not is a question for later, it just seems like the least painful way to manage edits and conversations.

dvdplm committed:
Add a SPEC.md

dvdplm committed:
Update SPEC.md

dvdplm committed:
Distinguish TSS/AKS needs. Start replication discussion.

dvdplm committed:
Resolve some TODOs, add more.
Describe new AKS node joining.

ameba23 commented on SPEC.md:
```diff
@@ -0,0 +1,233 @@
+
+*NOTE*: `secret_id`s should be unique and **random**, given that knowledge of a `secret_id` allows anyone to make remote API calls on behalf of the user.
+
+_TODO: Use correct endpoints, from the spec doc._
```
I have updated them above - the PUT and DELETE are not yet implemented so i left them as they are

dvdplm replied:
I think `deploy-api-key` must return a `secret_id` though right? And `make-request` take an `id` param so that a user can upload many different secrets yeah?

ameba23 commented on SPEC.md:
```diff
@@ -0,0 +1,233 @@
+
+_TODO: Use correct endpoints, from the spec doc._
+
+_TODO: Describe the ACL format and how it works in detail._
```
The access control is currently just a signature provided by the user. There is not (yet) way to allow multiple users to access the same API key.

And since we use sr25519, it is not currently possible to use a signature made by the entropy network and leverage programs to do access control. I'm not sure if we would want that anyway, as it will add several seconds onto the time it takes to get a response. Probably we are going to end up re-implementing some way of doing access control lists here.

ameba23 commented on SPEC.md:
```diff
@@ -0,0 +1,233 @@
+
+_TODO: Can users provide any `remote-url` like this or is it limited to a set of whitelisted URLs? Stored where? In the ACL? Or separate?_
+
+~~All calls to the AKS from the outside are signed and contain counter measures to replay attacks (likely a block number and/or a random nonce). The API responses are encrypted to the end user's public key before they are send back outside the AKS.~~
```
Currently we are using timestamp checks.  Which means we rely on client and sever to have accurate clocks (which these days is not so unreasonable to assume).

In `entropy-tss` we usually use block number, and we may well end up doing that here.  But we have had a lot of issues with automated tests and block number checks.

This is an argument for https - the TLS handshake means we don't need to worry about replay attacks.

> The API responses are encrypted to the end user's public key before they are send back outside the AKS.

Currently they are just sent clear.  Again, with https we would not need to implement this.

ameba23 commented on SPEC.md:
```diff
@@ -0,0 +1,233 @@
+
+_TODO: Spec the replay attack counter measures in detail. Given this all happens outside the chain, using the block number is pointless._
+
+**NOTE:** Users must do local key management and keep track of their `secret_id`s. A stolen `secret_id` can be used by anyone to execute remote API calls, so it's essentially equivalent to stealing the API key in the first place. That is an important matter worthy of its own separate discussion.
```
There are no secret ids - secrets are associated with public key of user plus url of service.

```suggestion
**NOTE:** Users must do local key management and keep track of their secret signing key. A stolen secret key can be used by anyone to execute remote API calls, so it's essentially equivalent to stealing the API key in the first place. That is an important matter worthy of its own separate discussion.
```

ameba23 commented on SPEC.md:
```diff
@@ -0,0 +1,233 @@
+To support the above scenario, the AKS nodes must keep the following data:
+
+- User secrets
+- User ACL, one for each remote rpc
```
User public keys and service hostnames

dvdplm committed:
Update SPEC.md

Co-authored-by: peg <ameba23@systemli.org>

dvdplm committed:
Update SPEC.md

Co-authored-by: peg <ameba23@systemli.org>

dvdplm committed:
Resolve some todos

dvdplm committed:
Incorporate Peg's state propagation ideas
Move meeting notes to bottom

dvdplm committed:
Moar edits

ameba23 commented on SPEC.md:
```diff
@@ -106,48 +106,16 @@ A rebooting AKS node is not too different from a new node. It must re-register o
-### SGX Sealing API as a keyserver for AKS/TSS
-
-–––> END Meeting notes, June 4th, 2025
+_TODO: Not convinced that the above is actually the simplest possible protocol that can work!_
```
Agree, its definitely not the simplest.

dvdplm committed:
Proof reading sessh

dvdplm committed:
AKS replicates

dvdplm committed:
Add reqs from #1398
Add placeholder links to issues for
- ACLs
- Payload encryption and replay attack counter measures

ameba23 committed:
Edit spec

ameba23 committed:
Merge dp-spec-doc

ameba23 committed:
Minor edit

ameba23 committed:
Add link to stub issue

ameba23 committed:
Add link to issue sealing

dvdplm committed:
Update SPEC.md

dvdplm committed:
Update SPEC.md

dvdplm committed:
Update SPEC.md

dvdplm committed:
Update SPEC.md

dvdplm committed:
Merge pull request #36 from entropyxyz/peg/changes-to-spec-doc00

Edits to spec doc

dvdplm committed:
Update SPEC.md
